src/Features/CSharp/Portable/RemoveAsyncModifier/CSharpRemoveAsyncModifierCodeFixProvider.cs

```diff
@@ -0,0 +1,118 @@
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp.ConvertSwitchStatementToExpression;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Formatting;
+using Microsoft.CodeAnalysis.CSharp.MakeMethodSynchronous;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.RemoveAsyncModifier;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using static Microsoft.CodeAnalysis.MakeMethodAsynchronous.AbstractMakeMethodAsynchronousCodeFixProvider;
+
+namespace Microsoft.CodeAnalysis.CSharp.RemoveAsyncModifier
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp, Name = PredefinedCodeFixProviderNames.MakeMethodSynchronous), Shared]
+    [ExtensionOrder(After = PredefinedCodeFixProviderNames.AddImport)]
+    internal partial class CSharpRemoveAsyncModifierCodeFixProvider : AbstractRemoveAsyncModifierCodeFixProvider
+    {
+        private const string CS1998 = nameof(CS1998); // This async method lacks 'await' operators and will run synchronously.
+
+        [ImportingConstructor]
+        [SuppressMessage("RoslynDiagnosticsReliability", "RS0033:Importing constructor should be [Obsolete]", Justification = "Used in test code: https://github.com/dotnet/roslyn/issues/42814")]
+        public CSharpRemoveAsyncModifierCodeFixProvider()
+        {
+        }
+
+        public override ImmutableArray<string> FixableDiagnosticIds { get; } = ImmutableArray.Create(CS1998);
+
+        protected override bool IsAsyncSupportingFunctionSyntax(SyntaxNode node)
+            => node.IsAsyncSupportingFunctionSyntax();
+
+        protected override bool CanFix(KnownTypes knownTypes, IMethodSymbol methodSymbolOpt)
+        => !methodSymbolOpt.ReturnType.OriginalDefinition.Equals(knownTypes._iAsyncEnumerableOfTTypeOpt) &&
+           !methodSymbolOpt.ReturnType.OriginalDefinition.Equals(knownTypes._iAsyncEnumeratorOfTTypeOpt);
+
+        protected override SyntaxNode ChangeReturnStatements(SyntaxNode node, SyntaxGenerator generator, KnownTypes knownTypes)
+        {
+            var converter = new ConvertReturnsToTaskRewriter(generator, knownTypes._taskType);
+
+            return node switch
+            {
+                MethodDeclarationSyntax method => FixMethodReturns(method, converter),
+                LocalFunctionStatementSyntax localFunction => FixLocalFunctionReturns(localFunction, converter),
+                AnonymousFunctionExpressionSyntax method => FixAnonymousFunctionReturns(method, converter),
+                _ => node,
+            };
+        }
+
+        private static AnonymousFunctionExpressionSyntax FixAnonymousFunctionReturns(AnonymousFunctionExpressionSyntax lambda, ConvertReturnsToTaskRewriter converter)
+        {
+            if (lambda.Block != null)
+            {
+                return lambda.WithBlock((BlockSyntax)converter.Visit(lambda.Block));
```

Comment:
you repeat this pattern a lot (including the casts).  consider extracting a helper for this that is strongly typed (like `ConvertBlock` and just do the cast one time in that.